### PR TITLE
(don't merge, needs rebase) Adding coercions for Array[Array[_]]

### PR DESCRIPTION
**Pull request by scottfrazer**

### CODE REVIEW

File: src/main/scala/cromwell/binding/command/ParameterCommandPart.scala [hunk 1]
```diff
@@ -31,7 +31,7 @@ case class ParameterCommandPart(attributes: Map[String, String], expression: Wdl
           /* Allow an expression to fail evaluation if one of the variables that it requires is optional (the type has ? after it, e.g. String?) */
           case Some(d) if d.postfixQuantifier.contains("?") => if (attributes.contains("default")) WdlString(attributes.get("default").head) else WdlString("")
           case Some(d) => throw new UnsupportedOperationException(s"Parameter ${v.variable} is required, but no value is specified")
-          case None => throw new UnsupportedOperationException(s"This should not happen: could not find declaration for ${v.variable}")
+          case None => throw new UnsupportedOperationException(s"ERROR: could not find declaration for ${v.variable}")
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/command/ParameterCommandPart.scala]: While ERROR is better than "This should not happen", isn't that kinda implied by being an exception?

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/command/ParameterCommandPart.scala]: Fair enough

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 2]
```diff
@@ -20,6 +21,15 @@ case class WdlArrayType(memberType: WdlType) extends WdlType {
       WdlArray(WdlArrayType(WdlFileType), wdlArray.value.map(str => WdlFile(str.asInstanceOf[WdlString].value)).toList)
     case wdlArray: WdlArray if wdlArray.wdlType.memberType == memberType => wdlArray
     case wdlArray: WdlArray if wdlArray.wdlType.memberType == WdlAnyType => coerceIterable(wdlArray.value)
+    case wdlArray: WdlArray if wdlArray.wdlType.memberType.isInstanceOf[WdlArrayType] && memberType.isInstanceOf[WdlArrayType] =>
+      val coercedElements = wdlArray.value.map(v => memberType.coerceRawValue(v))
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: does this need the `v => `?

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 3]
```diff
@@ -20,6 +21,15 @@ case class WdlArrayType(memberType: WdlType) extends WdlType {
...
+      val coercedArray = coercedElements.collect({case x: Failure[_] => x}) match {
+        case failures if failures.nonEmpty =>
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: I can't remember offhand but there's probably something in scalaz (which we use now) to handle this via an applicative, which is effectively what your'e doing here.

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 4]
```diff
@@ -20,6 +21,15 @@ case class WdlArrayType(memberType: WdlType) extends WdlType {
...
+          throw new UnsupportedOperationException(failures.map(f => s"Could not coerce $wdlArray: ${f.exception.getMessage}").mkString("\n"))
+        case _ =>
+          coercedElements.map(_.get) // .get because they're all Success
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: Move this up a line

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 5]
```diff
@@ -20,6 +21,15 @@ case class WdlArrayType(memberType: WdlType) extends WdlType {
...
+      }
+      WdlArray(WdlArrayType(memberType), coercedArray)
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: newline

**Pull request closed without merging**